atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/TransactionOutcomeMetrics.java

```diff
@@ -0,0 +1,117 @@
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.impl;
+
+import java.util.Map;
+import java.util.function.Predicate;
+
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.logging.LoggingArgs;
+import com.palantir.atlasdb.util.MetricsManager;
+import com.palantir.tritium.metrics.registry.MetricName;
+
+/**
+ * Tracks various states of an AtlasDB transaction, and marks associated meters in the
+ * {@link com.palantir.tritium.metrics.registry.TaggedMetricRegistry} associated with the provided
+ * {@link MetricsManager}.
+ */
+public class TransactionOutcomeMetrics {
+    private static final String SUCCESSFUL_COMMIT = "successfulCommit";
+    private static final String FAILED_COMMIT = "failedCommit";
+    private static final String ABORT = "abort";
+    private static final String WRITE_WRITE_CONFLICT = "writeWriteConflict";
+    private static final String READ_WRITE_CONFLICT = "readWriteConflict";
+    private static final String LOCKS_EXPIRED = "locksExpired";
+    private static final String PUT_UNLESS_EXISTS_FAILED = "putUnlessExistsFailed";
+    private static final String ROLLBACK_OTHER = "rollbackOther";
+
+    @VisibleForTesting
+    final MetricsManager metricsManager;
+
+    private final Predicate<TableReference> safeForLogging;
+
+    @VisibleForTesting
+    TransactionOutcomeMetrics(MetricsManager metricsManager, Predicate<TableReference> safeForLogging) {
+        this.metricsManager = metricsManager;
+        this.safeForLogging = safeForLogging;
+    }
+
+    public static TransactionOutcomeMetrics create(MetricsManager metricsManager) {
+        return new TransactionOutcomeMetrics(metricsManager, LoggingArgs::isSafe);
+    }
+
+    public void markSuccessfulCommit() {
+        getMeter(SUCCESSFUL_COMMIT).mark();
+    }
+
+    public void markFailedCommit() {
+        getMeter(FAILED_COMMIT).mark();
+    }
+
+    public void markAbort() {
+        getMeter(ABORT).mark();
+    }
+
+    public void markWriteWriteConflict(TableReference tableReference) {
+        getMeterForTable(WRITE_WRITE_CONFLICT, tableReference).mark();
+    }
+
+    public void markReadWriteConflict(TableReference tableReference) {
+        getMeterForTable(READ_WRITE_CONFLICT, tableReference).mark();
+    }
+
+    public void markLocksExpired() {
+        getMeter(LOCKS_EXPIRED).mark();
+    }
+
+    public void markPutUnlessExistsFailed() {
```

Comment:
Note for other reviewers: I initially thought this should also be per-table, but I infer that this is the putUnlessExists to the `_transactions` table.

Reply:
Now that I've moved the string constants to be enumeration names, I can add some javadoc clarifying what the outcomes mean.